### PR TITLE
add the implementation of create or update just for diff

### PR DESCRIPTION
**Pull request by YiMysty**
Please have a look at our [guidelines for contributions](https://github.com/Microsoft/AppCenter-SDK-Android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

A few sentences describing the overall goals of the pull request.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 1]
```diff
@@ -429,21 +429,26 @@ public void onCallFailed(Exception e) {
...
+                        }
 
-                    @Override
-                    public void completeFuture(Exception e) {
-                        Storage.this.completeFuture(e, result);
-                    }
-                });
+                        @Override
+                        public void completeFuture(Exception e) {
+                            Storage.this.completeFuture(e, result);
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: You don't need `Storage.this` I think here.

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/BaseOptions.java [hunk 2]
```diff
@@ -45,11 +45,10 @@ public int getDeviceTimeToLive() {
     }
 
     /**
-     * @param lastModified timestamp of when the document was last modified
-     * @return whether a document with the specified TTL has expired since it was last modified
+     * @param expiredAt timestamp of when the document was expired.
+     * @return whether a document has been expired.
```
Comment (dhei) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/BaseOptions.java]: wording "whether a document is expired."

Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/models/BaseOptions.java]: fixed~

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 3]
```diff
@@ -428,21 +428,26 @@ public void onCallFailed(Exception e) {
...
+                    result,
+                    new TokenExchangeServiceCallback() {
 
-                    @Override
-                    public void callCosmosDb(final TokenResult tokenResult) {
-                        callCosmosDbCreateOrUpdateApi(tokenResult, document, documentType, partition, documentId, writeOptions, result);
-                    }
+                        @Override
+                        public void callCosmosDb(final TokenResult tokenResult) {
+                            callCosmosDbCreateOrUpdateApi(tokenResult, document, documentType, partition, documentId, writeOptions, result);
```
Comment (YiMysty) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: if create succeed, should create the document in local storage too? with operation NIL?

Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: or leave it to create when read?

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java [hunk 4]
```diff
@@ -428,21 +428,26 @@ public void onCallFailed(Exception e) {
...
+                    partition,
+                    result,
+                    new TokenExchangeServiceCallback() {
 
-                    @Override
-                    public void callCosmosDb(final TokenResult tokenResult) {
-                        callCosmosDbCreateOrUpdateApi(tokenResult, document, documentType, partition, documentId, writeOptions, result);
-                    }
+                        @Override
+                        public void callCosmosDb(final TokenResult tokenResult) {
```
Comment (guperrot) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]:
```suggestion
                        public void callCosmosDb(TokenResult tokenResult) {
```

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java (hunk 3 above)
Comment (guperrot) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: We should cache when created yes, let's do it in a separate PR as this one is reviewed.

Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java]: will deal in the next pr 🙈

File: sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/LocalDocumentStorage.java [hunk 5]
```diff
@@ -150,6 +153,26 @@ private LocalDocumentStorage(DatabaseManager databaseManager) {
         return new Document<>(new StorageException("Document was not found in the cache."));
     }
 
+    <T> Document<T> createOrUpdate(String partition, String documentId, T document, Class<T> documentType, WriteOptions writeOptions){
```
Comment (t-nafle) [inline sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/LocalDocumentStorage.java]: You might want to run the formatter 👍

Comment (YiMysty) [reply sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/LocalDocumentStorage.java]: sure! will fix it :D

**Pull request merged**